### ISSUE TITLE
dual host setup

### ISSUE THREAD

**Issue by rainbowdn** [eclipse-cyclonedds/cyclonedds#435]
hi,
I want to make 2 hosts with different ip addresses to communicate via cyclone, The readme of the main page just explains how to setup dds is a single host. Could somebody please pointout a guid for setting up DDS in a distributed way so that my client and server can be executed on 2 different hosts? would hosts need to be on the same LAN?

**Comment by eboasson** [comment]
Hi @rainbowdn, the README explains so little because normally you don't have to do anything to make it work between hosts in a network: UDP/IP multicast is used for fully automatic discovery. The protocol starts out by multicasting to 239.255.0.1 at port 7400 [*] and any node that receives this message will start communicating.

Multicast is pretty widely available, but not literally everywhere. It also often don't play nicely with routers, so if you can stick to a single LAN it is definitely easiest.

If you really don't have multicast you can fall back to a unicast-only mode, but then you have to configure a list of peers. That'd be something like (beware of typos!):
```XML
<?xml version="1.0" encoding="UTF-8" ?>
<CycloneDDS xmlns="https://cdds.io/config" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="https://cdds.io/config https://raw.githubusercontent.com/eclipse-cyclonedds/cyclonedds/master/etc/cyclonedds.xsd">
    <Domain id="any">
        <General>
            <AllowMulticast>false</AllowMulticast>
        </General>
        <Discovery>
          <ParticipantIndex>auto</ParticipantIndex>
          <Peers>
            <Peer address="1.2.3.4"/>
            <Peer address="5.6.7.8"/>
          </Peers>
        </Discovery>
    </Domain>
</CycloneDDS>
```

The "ParticipantIndex" here forces it to use well-known port numbers (7410, 7412, 7414, ...) instead of kernel-allocated random port numbers. The discovery information in the multicast messages contains enough information not to have to use those well-known ones, but if you can't use multicast, you really need to know exactly which addresses to ping.

[*] Well, it would be a different port number if you change the domain id, and everything's configurable anyway.

**Comment by rainbowdn** [comment]
Thanks @eboasson , I assume 239.255.0.1 is also configurable, right? (cloud you please pointout where is the config file-is it configured in pom file? or is it a config file in dds setup directory- or if there is a manual that explains this) how often does the discovery via boradcast/multicast take place? is it only the first time or does it happen frequently (in this case, is it configurable via same file above?)

**Comment by eboasson** [comment]
@rainbowdn yes, that one is also configurable (239.255.0.1 is simply the default multicast address defined the specification). Every setting has a default and so there is actually no configuration file unless you create one yourself and set the environment variable ``CYCLONEDDS_URI`` to the file name. (It has a few more tricks up its sleeve: it is actually a comma-separate list of files of configuration files and XML configuration fragments.)

There is some documentation on this, the README gives a bit and provides (near the very end of the README) a pair of links to an overview of how all this works and an exhaustive list of options — see, e.g., https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/docs/manual/config.rst#discovery-of-participants--endpoints, https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/docs/manual/config.rst#discovery-addresses and https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/docs/manual/config.rst#timing-of-spdp-packets. A fancy XML editor (like VS Code with RedHat's XML extension) will grok the XSD file referenced in the README and provide you with tooltips, auto-complete, &c.

**Comment by thijsmie** [comment]
The question was answered and no further comments on problems were received, so I am assuming that everything is working now!
